### ISSUE TITLE
parallelize fmm

### ISSUE THREAD

**Issue by GaZ3ll3** [GaZ3ll3/femex#63]
currently fmm are running serially. Now there are two options to do it.
- cilkplus. supported by gcc-5/g++-5 and icc/icpc, easy to program and spawn new task.
- tbb.  
  - upward pass operates on source, bottom-up, on each level, the tasks are independent, which can run in parallel.
  - downward pass operates on tree, top-bottom, on each level, the tasks are also independent. 

The concurrent queue should be a good choice. 
- each task may add new tasks append on end
- when meeting current task, it should have all information needed from previous tasks.

**Comment by GaZ3ll3** [comment]
- [x] tree initialization parallelled.

**Comment by GaZ3ll3** [comment]
done with openmp.
